DT_INST_STRING_TOKEN_BY_IDX_OR macro

KozhinovAlexander opened the pull request:
This commit introduces the DT_INST_STRING_TOKEN_BY_IDX_OR macro, which allows specifying a default value if the indexed string token is not defined.

How-To-Test:

Introduce new property in any .yml file of your choice:

```yml
my-prop:
    required: true
    type: string-array
    enum:
      - FOO
      - BAR
```

Add it to any .dts:

```dts
node {
	my-prop = "FOO", "BAR";
	status = "okay";
};
```

Try to access from .c code an existing index:

```c
#define DT_INST_GET_MY_PROP(inst) DT_INST_STRING_TOKEN_BY_IDX_OR(inst, my_prop, 0, BAZ))
```

It shall return `FOO`.

Try to access from .c code a NON-existing index:

```c
#define DT_INST_GET_MY_PROP(inst) DT_INST_STRING_TOKEN_BY_IDX_OR(inst, my_prop, 3, BAZ))
```

It shall return `BAZ`.

UPD: Added tests under `tests/lib/devicetree/api`, thnx @JarmouniA

Run Twister Tests:

```bash
west twister -T tests/lib/devicetree/api
```

KozhinovAlexander committed:
zephyr: devicetree: add string token helper macros

This commit introduces the DT_INST_STRING_TOKEN_BY_IDX_OR()
and DT_STRING_TOKEN_BY_IDX_OR() macros, which allow
specifying a default value if the indexed string token is not
defined.

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

KozhinovAlexander commented:
@nashif I think - there is no reaction from assignee, please take a look.

etienne-lms reviewed (commented):
For commit "zephyr: devicetree: add new macros", suggestions for the commit message.

For the header line:
```
zephyr: devicetree: add DT_[INST_]STRING_TOKEN_BY_IDX_OR() macros
```
or
```
zephyr: devicetree: add string token helper macros
```

For the message body:
```patch
-This commit introduces the DT_INST_STRING_TOKEN_BY_IDX_OR
+This commit introduces the DT_INST_STRING_TOKEN_BY_IDX_OR()
-and DT_INST_STRING_TOKEN_BY_IDX macros, which allows
+and DT_STRING_TOKEN_BY_IDX_OR() macros, which allow
 specifying a default value if the indexed string token is not
 defined.

KozhinovAlexander commented:
@etienne-lms Thank you for your careful review. Every your suggestion was applied.

mathieuchopstm commented on include/zephyr/devicetree.h:
```diff
@@ -4418,6 +4432,18 @@
 	DT_STRING_TOKEN_BY_IDX(DT_DRV_INST(inst), prop, idx)
 
+/**
+ * @brief Like DT_INST_STRING_TOKEN_BY_IDX(), but with a fallback to default_value
```
```suggestion
 * @brief Like DT_INST_STRING_TOKEN_BY_IDX(), but with a fallback to @p default_value
```

etienne-lms replied:
My mistake. `@p` should be used when referencing an argument.

mathieuchopstm commented on include/zephyr/devicetree.h:
```diff
@@ -1322,6 +1323,19 @@
 	DT_CAT6(node_id, _P_, prop, _IDX_, idx, _STRING_TOKEN)
 
+/**
+ * @brief Like DT_STRING_TOKEN_BY_IDX(), but with a fallback to default_value
```
```suggestion
 * @brief Like DT_STRING_TOKEN_BY_IDX(), but with a fallback to @p default_value
```

mathieuchopstm commented on include/zephyr/devicetree.h:
```diff
@@ -1322,6 +1323,19 @@
+ * @param idx the index to get
+ * @param default_value a fallback value to expand to
+ * @return the element in @p prop at index @p idx as a token if indexed property
+ *         is present, @p default_value otherwise
```
Following existing conventions:
```suggestion
 * @return the element in @p prop at index @p idx as a token, or @p default_value
```

mathieuchopstm commented on include/zephyr/devicetree.h:
```diff
@@ -4418,6 +4432,18 @@
+ * @param idx the index to get
+ * @param default_value a fallback value to expand to
+ * @return the element in @p prop at index @p idx as a token if indexed property
+ *         is present, @p default_value otherwise
```
Following existing conventions:
```suggestion
 * @return the element in @p prop at index @p idx as a token, or @p default_value
```

KozhinovAlexander commented:
@mathieuchopstm I need you to take a look again especially regarding this two posts from me:
- Please find a single decision rearding this one: https://github.com/zephyrproject-rtos/zephyr/pull/97404#discussion_r2519170522
- Please let keep current implemtntation or provide a link with contribution rules if such exists: https://github.com/zephyrproject-rtos/zephyr/pull/97404#discussion_r2519155436

KozhinovAlexander commented on include/zephyr/devicetree.h:
```diff
@@ -2,6 +2,7 @@
  * SPDX-License-Identifier: Apache-2.0
  * Copyright (c) 2020 Nordic Semiconductor
  * Copyright (c) 2020, Linaro Ltd.
+ * Copyright (c) 2025 Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>
```
copyright were removed and stays now unchanged in this PR.

etienne-lms reviewed (dismissed):
LGTM. Thanks for the update.

KozhinovAlexander commented:
@etienne-lms Thank you for rocket fast review 😃

mbolivar commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -3503,6 +3514,13 @@ ZTEST(devicetree_api, test_string_idx_token)
 	zassert_equal(STRING_TOKEN_BY_IDX_VAR(DT_NODELABEL(test_str_array_token_1))[2],
 			token_second_idx_two, "");
 
+	/* Index is in range instance #0 corresponds to test_str_array_token_0 */
```
No, instance 0 does not have to be test_str_array_token_0. You are relying on whatever happened to work in your build environment today. This is a common error -- please read the doxygen documentation for DT_INST() carefully and rewrite this without making any assumptions not guaranteed by the API.

mbolivar reviewed (changes requested):
Thanks for the patch! There is a faulty assumption in the tests, though.

In addition to fixing that, can you please update the release notes file to mention the new macros?

Other than those issues, I don't see any blockers.

KozhinovAlexander committed:
tests: lib: devicetree: api: add new macro tests

add tests for DT_STRING_TOKEN_BY_IDX_OR macro
add tests for DT_INST_STRING_TOKEN_BY_IDX_OR macro

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

etienne-lms reviewed (commented):
LGTM with a minor comment.
